### ISSUE TITLE
Error in InputQueueingStateSaver/barrier

### ISSUE THREAD

**Issue by rubenvereecken** [tensorflow/tensorflow#11613]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Mint 18
- **TensorFlow installed from (source or binary)**: Binary (pip)
- **TensorFlow version (use command below)**: v1.2.0-rc2-21-g12f033d 1.2.0
- **Bazel version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: N/A

### Describe the problem
I encounter an error related to barriers when using `tf.contrib.training.batch_sequences_with_states`. I use this function to batch sequences. For now, just batches of 4. I do not understand the error it but I tried empirically poking at it and it only occurs with some capacity values. For example, it does not occur within 15 minutes with a `capacity` value of 16 (good for 4 batches) and cycles through the input sequences as expected. It does occur within a minute with `capacity` values of 32 and 64.

I know this is incredibly vague but I can explore more if you can help me understand the underlying issue. I do believe this question does not fit the StackOverflow format as this is buggy behaviour. 

### Source code / logs
Below the error log. One thing I do not understand for example is that it always complains about components 0-8. I have verified that this occurs for multiple keys. My best guess is that this has to do with `save_state` trying to save over state that has already been saved?

```
2017-07-19 17:16:52.465021: W tensorflow/core/framework/op_kernel.cc:1158] Invalid argument: Key 00030_of_00112:20120305_seq1 already has a value for component 8 in barrier _3_input/batch_seq_with_states/InputQueueingStateSaver/barrier
2017-07-19 17:16:52.465073: W tensorflow/core/kernels/queue_base.cc:294] _2_input/batch_seq_with_states/fifo_queue: Skipping cancelled enqueue attempt with queue not closed
2017-07-19 17:16:52.465117: W tensorflow/core/framework/op_kernel.cc:1158] Invalid argument: Key 00030_of_00112:20120305_seq1 already has a value for component 7 in barrier _3_input/batch_seq_with_states/InputQueueingStateSaver/barrier
2017-07-19 17:16:52.465168: W tensorflow/core/framework/op_kernel.cc:1158] Invalid argument: Key 00030_of_00112:20120305_seq1 already has a value for component 5 in barrier _3_input/batch_seq_with_states/InputQueueingStateSaver/barrier
2017-07-19 17:16:52.465226: W tensorflow/core/framework/op_kernel.cc:1158] Invalid argument: Key 00030_of_00112:20120305_seq1 already has a value for component 3 in barrier _3_input/batch_seq_with_states/InputQueueingStateSaver/barrier
2017-07-19 17:16:52.465261: W tensorflow/core/framework/op_kernel.cc:1158] Invalid argument: Key 00030_of_00112:20120305_seq1 already has a value for component 4 in barrier _3_input/batch_seq_with_states/InputQueueingStateSaver/barrier
2017-07-19 17:16:52.465292: W tensorflow/core/framework/op_kernel.cc:1158] Invalid argument: Key 00030_of_00112:20120305_seq1 already has a value for component 0 in barrier _3_input/batch_seq_with_states/InputQueueingStateSaver/barrier
2017-07-19 17:16:52.465324: W tensorflow/core/framework/op_kernel.cc:1158] Invalid argument: Key 00030_of_00112:20120305_seq1 already has a value for component 1 in barrier _3_input/batch_seq_with_states/InputQueueingStateSaver/barrier
2017-07-19 17:16:52.465798: W tensorflow/core/framework/op_kernel.cc:1158] Invalid argument: Key 00030_of_00112:20120305_seq1 already has a value for component 2 in barrier _3_input/batch_seq_with_states/InputQueueingStateSaver/barrier
2017-07-19 17:16:52.469856: W tensorflow/core/framework/op_kernel.cc:1158] Invalid argument: Key 00030_of_00112:20120305_seq1 already has a value for component 6 in barrier _3_input/batch_seq_with_states/InputQueueingStateSaver/barrier
INFO:tensorflow:Error reported to Coordinator: <class 'tensorflow.python.framework.errors_impl.InvalidArgumentError'>, Key 00030_of_00112:20120305_seq1 already has a value for component 8 in barrier _3_input/batch_seq_with_states/InputQueueingStateSaver/barrier
         [[Node: input/batch_seq_with_states/InputQueueingStateSaver/BarrierInsertContext_num_frames = BarrierInsertMany[T=DT_INT64, _class=["loc:@input/batch_seq_with_states/fifo_queue"], component_index=8, _device="/job:localhost/replica:0/task:0/cpu:0"](input/batch_seq_with_states/InputQueueingStateSaver/barrier, input/batch_seq_with_states/InputQueueingStateSaver/StringJoinCurrentKeys, input/batch_seq_with_states/InputQueueingStateSaver/dims_checked_expanded_context_num_frames)]]
17:16:52 [tensorflow          ] [INFO    ] : Error reported to Coordinator: <class 'tensorflow.python.framework.errors_impl.InvalidArgumentError'>, Key 00030_of_00112:20120305_seq1 already has a value for component 8 in barrier _3_input/batch_seq_with_states/InputQueueingStateSaver/barrier
         [[Node: input/batch_seq_with_states/InputQueueingStateSaver/BarrierInsertContext_num_frames = BarrierInsertMany[T=DT_INT64, _class=["loc:@input/batch_seq_with_states/fifo_queue"], component_index=8, _device="/job:localhost/replica:0/task:0/cpu:0"](input/batch_seq_with_states/InputQueueingStateSaver/barrier, input/batch_seq_with_states/InputQueueingStateSaver/StringJoinCurrentKeys, input/batch_seq_with_states/InputQueueingStateSaver/dims_checked_expanded_context_num_frames)]]
Traceback (most recent call last):
  File "./train.py", line 108, in <module>
    solver.train(**pick(all_args, list(inspect.signature(solver.train).parameters)))
  File "/home/ruben/attention/attend/attend/solver.py", line 122, in train
    coord.join(threads + input_threads)
  File "/home/ruben/anaconda3/lib/python3.6/site-packages/tensorflow/python/training/coordinator.py", line 389, in join
    six.reraise(*self._exc_info_to_raise)
  File "/home/ruben/anaconda3/lib/python3.6/site-packages/six.py", line 686, in reraise
    raise value
  File "/home/ruben/anaconda3/lib/python3.6/site-packages/tensorflow/python/training/queue_runner_impl.py", line 238, in _run
    enqueue_callable()
  File "/home/ruben/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1063, in _single_operation_run
    target_list_as_strings, status, None)
  File "/home/ruben/anaconda3/lib/python3.6/contextlib.py", line 89, in __exit__
    next(self.gen)
  File "/home/ruben/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/errors_impl.py", line 466, in raise_exception_on_not_ok_status
    pywrap_tensorflow.TF_GetCode(status))
tensorflow.python.framework.errors_impl.InvalidArgumentError: Key 00030_of_00112:20120305_seq1 already has a value for component 8 in barrier _3_input/batch_seq_with_states/InputQueueingStateSaver/barrier
         [[Node: input/batch_seq_with_states/InputQueueingStateSaver/BarrierInsertContext_num_frames = BarrierInsertMany[T=DT_INT64, _class=["loc:@input/batch_seq_with_states/fifo_queue"], component_index=8, _device="/job:localhost/replica:0/task:0/cpu:0"](input/batch_seq_with_states/InputQueueingStateSaver/barrier, input/batch_seq_with_states/InputQueueingStateSaver/StringJoinCurrentKeys, input/batch_seq_with_states/InputQueueingStateSaver/dims_checked_expanded_context_nu

```

Relevant source code in `provider.py`:

```python
batch = tf.contrib.training.batch_sequences_with_states(
        input_sequences={
            'images': example,
            self.feat_name: target,
        },
        input_key      = context['key'],
        input_context  = context,
        input_length   = tf.cast(context['num_frames'], tf.int32),
        initial_states = initial_states,
        num_unroll     = self.time_steps,
        batch_size     = self.batch_size,
        num_threads    = 2, # TODO change
        capacity       = self.batch_size * 4,
        name           = 'batch_seq_with_states'
        )


```

**Comment by rubenvereecken** [comment]
Additionally, I tested with a file holding exactly 3 examples, using a batch size of 1. A capacity of 1 would work. A capacity of 2 results in the same error.

**Comment by rubenvereecken** [comment]
I could see how this might occur if the same input key is encountered (by e.g. reading the same file twice). Could anyone confirm?

**Comment by raoconn** [comment]
I have encountered the same error, running TensorFlow 1.2.0 on centos 7 with 4 GeForce GTX 980 Ti GPUs .  I am experimenting with a number of different recurrent neural network designs using `tf.contrib.training.batch_sequences_with_states` and `tf.nn.static_state_saving_rnn` and get the InvalidArgumentError only for some of them.  For instance, the RNN could have one or two layers and it could use GRU or LSTM cells.  With one layer and GRU cells, I have run 500 epochs (which results in about 536000 steps) without getting the error.  With two layers and GRU cells or one layer and LSTM cells, running 500 epochs results in the Invalid Argument Error at some point along the way (usually after at least 100000 and up to 400000 steps, depending on other parameters that I am modifying such as the number of cells in each layer).  With two layers and LSTM cells, the InvalidArgumentError occurs almost immediately (within the first 100 steps) and happens even if I have the queue set to read the file only once (1 epoch).  I have used different input files (with similar data, all in tfrecords format) and encountered the same error regardless of the file.

Relevant code:
```
num_unroll = 100
batch_size = 100
num_layers = config.num_layers
size = config.size
use_lstm = config.use_lstm
def single_cell():
    """Sets the basic cell to be either LSTM or GRU."""
    if use_lstm:
        return tf.nn.rnn_cell.BasicLSTMCell(size)
    return tf.nn.rnn_cell.GRUCell(size)
if num_layers > 1:
    stacked_cells = tf.nn.rnn_cell.MultiRNNCell([single_cell() for _ in range(num_layers)])
else:
    stacked_cells = single_cell()
state_name = 'rnn_state'
if num_layers > 1 and use_lstm:
    initial_state = tf.zeros((stacked_cells.state_size[0][0],), dtype=data_type())
    state_names = ((state_name, state_name),) * num_layers
elif num_layers > 1:
    initial_state = tf.zeros((stacked_cells.state_size[0],), dtype=data_type())
    state_names = (state_name,) * num_layers
elif use_lstm:
    initial_state = tf.zeros((stacked_cells.state_size[0],), dtype=data_type())
    state_names = (state_name, state_name)
else:
    initial_state = tf.zeros((stacked_cells.state_size,), dtype=data_type())
    state_names = state_name
rnn_initial_state = {'rnn_state': initial_state}
batch = tf.contrib.training.batch_sequences_with_states(
    input_key=inputs[0]['key'],
    input_sequences=inputs[1],
    input_context=inputs[0],
    input_length=tf.cast(inputs[0]['length'], tf.int32),
    initial_states=rnn_initial_state,
    num_unroll=num_unroll,
    batch_size=batch_size,
    allow_small_batch=True,
    num_threads=1,
    capacity=batch_size*4,
    pad=True,
    make_keys_unique=True)
inputs = data = batch.sequences['alert_id']
labels = batch.context['label']
lengths = batch.length
embedder = tf.get_variable('embedder', [config.max_alert_id, config.embedding_size])
vectors = tf.nn.embedding_lookup([embedder], data, name='embed')
vectors = tf.unstack(
    value=vectors,
    num=num_unroll,
    axis=1)
outputs, state = tf.nn.static_state_saving_rnn(
    stacked_cells,
    vectors,
    state_saver=batch,
    state_name=state_names)
```
All of the parameters involved in the batching are unchanged across my experiments, except `initial_states` which has to change according to the number of layers and type of RNN cell used, but I don't see why that should affect the reading of the data.

I don't know if it's relevant, but the error for my data always corresponds to component 13.

**Comment by gongbudaizhe** [comment]
Does the value of the key matters to you?

The setting below works for me: 

```python
input_key=tf.py_func(lambda: uuid.uuid4().hex, [], tf.string),
```

**Comment by raoconn** [comment]
Setting
```
input_key=tf.py_func(lambda: uuid.uuid4().hex, [], tf.string),
``` 
still produces the error for me.

**Comment by ebrevdo** [comment]
Try adding the argument `make_keys_unique=True` to `batch_sequences_with_queues`.

**Comment by rubenvereecken** [comment]
@ebrevdo good call, it does make sure sequences are not repeated in the same batch and it seems to have fixed it for me. I do get a tonne of barrier warnings but they can be ignored.

**Comment by raoconn** [comment]
That argument was already set to true for me.

**Comment by axelv** [comment]
Try implementing a custom key generator. If you use the option `make_keys_unique=True` Tensorflow uses a random key generator with a 100 mil possibilities:
[sequence_queueing_state_saver.py](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/contrib/training/python/training/sequence_queueing_state_saver.py)
```    
if make_keys_unique:
      input_key = string_ops.string_join([
          input_key,
          string_ops.as_string(
              random_ops.random_uniform(
                  (), minval=0, maxval=100000000, dtype=dtypes.int32,
                  seed=make_keys_unique_seed))])
```
A counting key generator fixed it for me:
```
key = tf.Variable(0)
```
```
batch = tf.contrib.training.batch_sequences_with_states(
        input_key=tf.as_string(key.assign_add(1)), ...)
```
However, I don't think I need millions of unique keys in the queue. So probably there is another underlying problem in my model.

**Comment by raoconn** [comment]
The counting key generator did not solve the problem for me, nor did combining the counting key with `make_keys_unique=True`.  I did learn that the error is happening (when I run with LSTM cells, 2 layers, and `num_epochs=1`, anyway) on the 111th item, but I don't think that's relevant.

I continue to find it odd that changing parameters that have no effect on the arguments passed to `batch_sequences_with_states` (e.g. learning rate parameters) affect whether or not this error occurs.  Any given parameter set has a consistent effect across repeated attempts (down to the key, except for the part generated by `make_keys_unique`, and component information in the error message), but changing the parameter set changes when or even if the error is triggered.  (Or, at least, some parameter sets won't trigger it within 500 epochs.)

For the most part I can work around this because training occurs for some time before the error, and I can simply restore the last checkpoint and continue until the error happens again.  However, that doesn't actually solve the problem.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by ebrevdo** [comment]
@raoconn can you give a minimal reproducible failure snippet?  link to gist is fine, but it should be as small as possible.  your issue is hard to debug without that because it's so specific to your data and code.

**Comment by raoconn** [comment]
This is no longer relevant to me.  I appreciate the effort put in, but I do not believe that it's worth either my time or anyone else's to continue investigating this unless there are others who are still having problems.

**Comment by ebrevdo** [comment]
OK; feel free to reopen if this bites you again.

**Comment by hellozjj** [comment]
ops.....This problem bothers me
